src/rtc.rs

```diff
@@ -0,0 +1,448 @@
+                    .clear_bit()
+            });
+        });
+
+        Ok(())
+    }
+
+    fn set_seconds(&mut self, seconds: u8) -> Result<(), Self::Error> {
+        if seconds > 59 {
+            return Err(Error::InvalidInputData);
+        }
+        let (st, su) = bcd2_encode(seconds as u32);
+        self.modify(|regs| regs.tr.write(|w| unsafe { w.st().bits(st).su().bits(su) }));
+
+        Ok(())
+    }
+
+    fn set_minutes(&mut self, minutes: u8) -> Result<(), Self::Error> {
+        if minutes > 59 {
+            return Err(Error::InvalidInputData);
+        }
+        let (mnt, mnu) = bcd2_encode(minutes as u32);
+        self.modify(|regs| {
+            regs.tr
+                .write(|w| unsafe { w.mnt().bits(mnt).mnu().bits(mnu) })
+        });
+
+        Ok(())
+    }
+
+    fn set_hours(&mut self, hours: rtcc::Hours) -> Result<(), Self::Error> {
+        let (ht, hu) = hours_to_register(hours)?;
+        match hours {
+            Hours::H24(_h) => {
+                self.set_24h_fmt();
+                self.modify(|regs| regs.tr.write(|w| unsafe { w.ht().bits(ht).hu().bits(hu) }));
+            }
+            Hours::AM(_h) | Hours::PM(_h) => {
+                self.set_12h_fmt();
+                self.modify(|regs| regs.tr.write(|w| unsafe { w.ht().bits(ht).hu().bits(hu) }));
+            }
+        }
+
+        Ok(())
+    }
+
+    fn set_weekday(&mut self, weekday: u8) -> Result<(), Self::Error> {
+        if (weekday < 1) || (weekday > 7) {
+            return Err(Error::InvalidInputData);
+        }
+        self.modify(|regs| regs.dr.write(|w| unsafe { w.wdu().bits(weekday) }));
+
+        Ok(())
+    }
+
+    fn set_day(&mut self, day: u8) -> Result<(), Self::Error> {
+        if (day < 1) | (day > 31) {
+            return Err(Error::InvalidInputData);
+        }
+        let (dt, du) = bcd2_encode(day as u32);
+        self.modify(|regs| regs.dr.write(|w| unsafe { w.dt().bits(dt).du().bits(du) }));
+
+        Ok(())
+    }
+
+    fn set_month(&mut self, month: u8) -> Result<(), Self::Error> {
+        if (month < 1) | (month > 12) {
+            return Err(Error::InvalidInputData);
+        }
+        let (mt, mu) = bcd2_encode(month as u32);
+        self.modify(|regs| {
+            regs.dr
+                .write(|w| unsafe { w.mt().bit(mt > 0).mu().bits(mu) })
+        });
+
+        Ok(())
+    }
+
+    fn set_year(&mut self, year: u16) -> Result<(), Self::Error> {
+        if (year < 1970) | (year > 2038) {
+            return Err(Error::InvalidInputData);
+        }
+        let (yt, yu) = bcd2_encode(year as u32);
+        self.modify(|regs| regs.dr.write(|w| unsafe { w.yt().bits(yt).yu().bits(yu) }));
+
+        Ok(())
+    }
+
+    /// set date using NaiveDate (ISO 8601 calendar date without timezone)
+    /// WeekDay is set using set_weekday method
+    fn set_date(&mut self, date: &NaiveDate) -> Result<(), Self::Error> {
+        let (yt, yu) = bcd2_encode((date.year() - 1970) as u32);
```

Comment:
If `date.year()` is less than 1970, this will underflow. You should check that and return an `InvalidInputData` error in this case